File: 0202-happy-number/0202-happy-number.py
class Solution(object):
    def isHappy(self, n):
        # POA
        # 1) Understand what a Happy number is(DONE)
        # 2) implement a hashset to keep track of Happy Numbers that are used to avoid an infinite loop
        # 3) Code a function that will handle the math behind the squaring
        # 4) Do we ever visit a number twice?
        
        # STEP ONE create a HashSet of visited numbers when squarred and summed. 
        visited_num = set()
        
        # STEP TWO compute the sum of squares until a duplicate value is repeated
        
        while n not in visited_num:
            # add num to visited Hashset
            visited_num.add(n)
            # compute the sum of squares through a helper function 
            n = self.sumOfSquares(n)
            
            #  If we get a happy number.
            if n == 1:
                return True 
            
        # if the value was NOT one and was visited twice during the loop    
        return False 
     # Helper function to calculate the sum of Squares        
    def sumOfSquares(self, n):
        # TODO: Compute the square of digits
        
        # variable to store the output
        output = 0
        
        # sum the squares
        while n:
            # get the digit for the second square
            digit = n % 10
            # square the digit
            digit = digit ** 2
            # add the digit to output
            output+=digit
            # update n to find the next digit
            n = n // 10
            
        return output 
            
            
    
        
        
        
        
           
        

File: 0202-happy-number/0202-happy-number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    //STEP ONE Define a helper function to calculate the sum of the squares of digits
    const getDigitSquareSum = (num) => {
    // Initialize a variable to store the sum
    let sum = 0;

    // Continue the loop as long as num is greater than 0
    while (num > 0) {
        // Extract the last digit of the number
        const digit = num % 10;

        // Add the square of the digit to the sum
        sum += digit * digit;

        // Remove the last digit from the number
        num = Math.floor(num / 10);
    }

    // Return the final sum of the squares of digits
    return sum;
};
//     STEP TWO: create a HashSet and monitor the set
    // Use a HashSet to detect cycles
    const seen = new Set();
//     if the number isn't one and is NOT in the set.
    while (n !== 1 && !seen.has(n)) {
        // Add the current number to the set
        seen.add(n);
        // Calculate the sum of the squares of digits
        n = getDigitSquareSum(n);
    }

    // If n is 1, it's a happy number; otherwise, it's not
    return n === 1;
};

// Example usage:
console.log(isHappy(19)); // Output: true
console.log(isHappy(2));  // Output: false
